main.py

```diff
@@ -0,0 +1,160 @@
+#   - https://github.com/irods/python-irodsclient/blob/master/irods/ticket.py
+#   - https://groups.google.com/g/iROD-Chat/c/gVfxH1Nn3Xc/m/EB1YchTOAAAJ
+#   - https://github.com/irods/python-irodsclient/commit/efd1c7495f454242dce81f03e598741cfabd955b#diff-371634f3c7d0926eb7a5118ddef62e79ee9ef2b3e12f893cc49cbe4033e7d270
+class GenQueryTicket(Model):
+    id              = Column(Integer, 'TICKET_ID',        2200)
+    string          = Column(String,  'TICKET_STRING',    2201)
+    type            = Column(String,  'TICKET_TYPE',      2202)
+    collection_name = Column(String,  'TICKET_COLL_NAME', 2228)
+
+def get_username_and_password(auth_header):
+    # Remove the "Basic " prefix.
+    auth_header = request.headers.get('authorization')[6:]
+    app.logger.info('auth header = [%s]', auth_header)
+
+    username, password = base64.b64decode(auth_header).decode('utf-8').split(':')
+    app.logger.info('username = [%s], password = [%s]', username, password)
+
+    return username, password
+
+def make_irods_credentials_dict(username, password):
+    return {
+        'host': 'localhost',
+        'port': 1247,
+        'user': username,
+        'password': password,
+        'zone': 'tempZone'
+    }
+
+def generate_token(data):
+    #key = jwk.JWK(generate='oct', size=256)
+    key = jwk.JWK.from_password(IRODS_TICKET_BOOTH_SHARED_SECRET)
+
+    token = jwt.JWT(header={'alg': 'HS256'}, claims=data)
+    token.make_signed_token(key)
+    return token.serialize()
+
+    # An encrypted token. This may be helpful in the final implementation.
+    #etoken = jwt.JWT(header={'alg': 'A256KW', 'enc': 'A256CBC-HS512'}, claims=token.serialize())
+    #etoken.make_encrypted_token(key)
+    #return etoken.serialize()
+
+@app.route("/create")
+def create():
+    if 'authorization' not in request.headers:
+        return 'Authorization header is not set.'
+
+    username, password = get_username_and_password(request.headers.get('authorization'))
+
+    if 'collection' not in request.args:
+        return '"collection" argument is not set.'
+
+    # Verify that the collection exists.
+    collection = Markup(request.args.get('collection', '')).unescape()
+    app.logger.info('collection = [%s]', collection)
+
+    # Permissions default to 'read'.
+    permission = 'read'
+    if 'permission' in request.args:
+        permission = Markup(request.args.get('permission', '')).unescape()
+        if permission not in ['read', 'write']:
+            return 'Invalid ticket permission [{}]'.format(permission)
+
+    with iRODSSession(**make_irods_credentials_dict(username, password)) as session:
+        if not session.collections.exists(collection):
+            return "Collection does not exist or user does not have permission to access the collection"
+
+        # Generate a ticket with the requested permissions for a collection.
+        # TODO What happens if the ticket matches a preexisting token?
```

Comment:
I think there is a unique constraint in the catalog for the ticket value itself.   Is that what you're asking about?

Reply:
ooh, i see, probably should try 2-3x and then stop and give an error.